### PR TITLE
TyDi QA Demo

### PR DESCRIPTION
**Pull request by aryan1107**
A finetuned TyDiQA model demo (Uses Huggingface & Jax version):

Screen Shot 2022-07-19 at 4 47 38 PM

### CODE REVIEW

File: lit_nlp/examples/models/tydi.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  }
+  @property
+  def max_seq_length(self):
+    return self.model.config.max_position_embeddings
+
+  def __init__(self, 
+              model_name="mrm8488/bert-multi-cased-finedtuned-xquad-tydiqa-goldp", 
+              model=FlaxBertForQuestionAnswering.from_pretrained("mrm8488/bert-multi-cased-finedtuned-xquad-tydiqa-goldp"),
+              tokenizer=BertTokenizer.from_pretrained("mrm8488/bert-multi-cased-finedtuned-xquad-tydiqa-goldp"),
+              **config_kw):
```
Comment (RyanMullins) [inline lit_nlp/examples/models/tydi.py:101]:
```suggestion
  def __init__(self, 
              model_name="mrm8488/bert-multi-cased-finedtuned-xquad-tydiqa-goldp", 
              model=None,
              tokenizer=None,
              **config_kw):
```

File: lit_nlp/examples/models/tydi.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    super().__init__()
+    self.config = T5ModelConfig(**config_kw)
+    self.tokenizer = tokenizer or transformers.AutoTokenizer.from_pretrained(
+        model_name, use_fast=False)
```
Comment (RyanMullins) [inline lit_nlp/examples/models/tydi.py:105]:
```suggestion
    self.tokenizer = tokenizer or BertTokenizer.from_pretrained(model_name)
```

File: lit_nlp/examples/models/tydi.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    # TODO(lit-dev): switch to TFBertForPreTraining to get the next-sentence
+    # prediction head as well.
+    self.model = model or model_utils.load_pretrained(
+        transformers.TFBertForMaskedLM,
+        model_name,
+        output_hidden_states=True,
+        output_attentions=True)
```
Comment (RyanMullins) [inline lit_nlp/examples/models/tydi.py:112]:
```suggestion
    self.model = model or FlaxBertForQuestionAnswering.from_pretrained(model_name)
```

File: lit_nlp/examples/models/tydi.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+import tensorflow as tf
+# tensorflow_text is required for T5 SavedModel
+import tensorflow_text  # pylint: disable=unused-import
+import transformers
+from transformers import BertTokenizer
+from transformers import FlaxBertForQuestionAnswering
+
+from rouge_score import rouge_scorer
+
+JsonDict = lit_types.JsonDict
```
Comment (RyanMullins) [inline lit_nlp/examples/models/tydi.py:20]:
```suggestion
import transformers

from rouge_score import rouge_scorer

BertTokenizer = transformers.BertTokenizer
FlaxBertForQuestionAnswering = transformers.FlaxBertForQuestionAnswering
JsonDict = lit_types.JsonDict
```

This is how we get around [the import restrictions](https://google.github.io/styleguide/pyguide.html#22-imports) in Google's Python Style Guide. Since we can't import the classes directly, and writing `transformers.{ClassName}` is less readable than we might want, we create local variables with the same names as the classes we want and assign them the value from the relevant modules.

File: lit_nlp/examples/models/tydi.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+  # Check outputs
+  ospec = model.output_spec()
+  assert "output_text" in ospec
+  assert isinstance(
+      ospec["output_text"],
+      (lit_types.GeneratedText, lit_types.GeneratedTextCandidates))
+  assert ospec["output_text"].parent == "target_text"
+
+  return model
```
Comment (RyanMullins) [inline lit_nlp/examples/models/tydi.py:83]: Import these from the T5 module using the pattern mentioned above for getting around [Python Style Guide's import restrictions](https://google.github.io/styleguide/pyguide.html#22-imports).

```python
import ./t5 as t5

T5ModelConfig = t5.T5ModelConfig

masked_token_mean = t5.masked_token_mean
validate_t5_model = t5.validate_t5_model
```

Also, are we sure we even need these since this is a Flax model?

File: lit_nlp/examples/models/tydi.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    question = []
+    for i in inputs:
+        question.append(i['question'])
+        context.append(i['context'])
+
+    prediction_output = []
+    for i in range(len(inputs)):
+        model = FlaxBertForQuestionAnswering.from_pretrained("mrm8488/bert-multi-cased-finedtuned-xquad-tydiqa-goldp")
+        inputs = self.tokenizer(question[i], context[i], return_tensors="jax",padding=True)
+        outputs = model(**inputs)
```
Comment (RyanMullins) [inline lit_nlp/examples/models/tydi.py:144]:
```suggestion
        inputs = self.tokenizer(question[i], context[i], return_tensors="jax", padding=True)
        outputs = self.model(**inputs)
```

File: lit_nlp/examples/models/tydi.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        "context": lit_types.TextSegment(),
+        "question": lit_types.TextSegment(required=False),
+    }
+
+  def output_spec(self):
+    spec = spec = {
+        "output_text": lit_types.GeneratedText(parent="question")
+    }
+    spec["rougeL"] = lit_types.Scalar()
+    return spec
```
Comment (RyanMullins) [inline lit_nlp/examples/models/tydi.py:176]:
```suggestion
    return {
        "output_text": lit_types.GeneratedText(parent="question"),
        "rougeL": lit_types.Scalar()
    }
```

File: environment.yml [hunk 8]
```diff
@@ -32,7 +32,7 @@ dependencies:
         - sacrebleu
         - numba==0.53.1
         - umap-learn==0.5.1
-        - transformers==4.1.1
```
[the remark this answers is not in the document]
Comment (RyanMullins) [reply environment.yml:35]: Answered my own question here... `FlaxBertForQuestionAnswering` was added to `transformers` in v4.5.0.

File: lit_nlp/examples/models/tydi.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    convert it to jax. Jax output is not same it gives only:
+    start_logits, end_logits, hiden stages and attentions
+
+    Args:
+      encoded_inputs: Dict as returned from Tokenizer for inputs.
+
+    Returns:
+      batched_outputs: Dict[str, jax]
+    """
+    m = FlaxBertForQuestionAnswering.from_pretrained("mrm8488/bert-multi-cased-finedtuned-xquad-tydiqa-goldp")
```
[the remark this answers is not in the document]
Comment (aryan1107) [reply lit_nlp/examples/models/tydi.py]: Yes when I was using self.model I kept getting this error: `'str' object is not callable`

Screen Shot 2022-06-15 at 2 01 08 PM

File: lit_nlp/examples/tydi_demo.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  for model_name_or_path in _MODELS.value:
+    # Ignore path prefix, if using /path/to/<model_name> to load from a
+    # specific directory rather than the default shortcut.
+    model_name = os.path.basename(model_name_or_path)
+    models[model_name] = tydi.TydiModel(
+          model=model_name_or_path,
+          model_name=model_name_or_path,
+          num_to_generate=_NUM_TO_GEN.value,
+          token_top_k=_TOKEN_TOP_K.value,
+          output_attention=False)
```
Comment (RyanMullins) [inline lit_nlp/examples/tydi_demo.py:80]: Figured out a few problems originating here (and also with Python just generally being Python)...

We were seeing bugs in the `TyDiModel.predict_minibatch(...)` function when [using `self.model`](https://github.com/PAIR-code/lit/pull/753/files#diff-e66e2ed65042046b3905d9bc4b44fa2c3bb19cb8c07da5a9540002725714fab0R144) because the `model` parameter in this initializer was being passed a string value when it should either be `None` or a `FlaxBertForQuestionAnswering` instance. 

Looking at the [docs for `FlaxBertForQuestionAnswering`](https://huggingface.co/transformers/v4.6.0/model_doc/bert.html#flaxbertforquestionanswering), it also doesn't accept the config parameters being passed in here (e.g., `num_to_generate`, _token_top_k`, `output_attention`), so we can probably simplify this initialization to the following.

```suggestion
    models[model_name] = tydi.TydiModel(model_name=model_name_or_path)
```

I have a few more comments in the `tydi.py` file above related to actually getting predictions from the model.

File: lit_nlp/examples/models/tydi.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+        output = {
+            "output_text" : self.tokenizer.decode(predict_answer_tokens)
+        }
+        prediction_output.append(output)
+    
+    print('Predictions for the data is---->/n')
+    print(prediction_output)
+
+    # returning list of prediction
+    return prediction_output
```
Comment (RyanMullins) [inline lit_nlp/examples/models/tydi.py:162]: Looking over the example provided with the [`FlaxBertForQuestionAnswering.__call__(...)` docs](https://huggingface.co/transformers/v4.6.0/model_doc/bert.html#transformers.FlaxBertForQuestionAnswering.__call__), I think we need to change this implementation a bit... 
```suggestion
  def predict_minibatch(self, inputs):
    """Predict on a single minibatch of examples."""
    # tokenize the text. -> then return prediction
  
    # Text as sequence of sentencepiece ID"s.
    prediction_output = []
    for inp in inputs:
      tokenized = self.tokenizer(inp['question'], inp['context'], return_tensors="jax", padding=True)
      output = self.model(**tokenized)
      answer_start_index = output.start_logits.argmax()
      answer_end_index = output.end_logits.argmax()
      predict_answer_tokens = inputs.input_ids[0, answer_start_index : answer_end_index + 1]
      prediction_output.append({
          "output_text" : self.tokenizer.decode(predict_answer_tokens)
      })
    
    print('Predictions for the data is---->/n')
    print(prediction_output)

    # returning list of prediction
    return prediction_output
```

**Comment by RyanMullins** [comment]
# Notes from 23 Jun 2022 meeting

* Explore using T5X ([GitHub](https://github.com/google-research/t5x), [PyPi](https://pypi.org/project/t5x/)) instead of Transformers for loading the Flax model. 
    * This is used internally at Google
    * Configuration is done in [Gin](https://github.com/google/gin-config)
    * README does not have info on loading saved Flax models, look into the [full docs](https://github.com/google-research/t5x/blob/main/docs/index.md) for that. 

## Action Items

- [ ] @RyanMullins to do code review today
- [ ] @RyanMullins to ask John Clark about a T5X model for TyDi QA we could use instead of the current model
- [ ] @aryan1107 Add an exact match metric
- [ ] @aryan1107 Do a dataset for each language
- [ ] @aryan1107 Port modeling code to T5X

**Comment by aryan1107** [comment]
Adding branch links which may be helpful for review:

[TyDi demo](https://github.com/aryan1107/lit/tree/dev)
[dalle-mini-demo](https://github.com/aryan1107/lit/tree/dev-dalle-demo)

File: lit_nlp/examples/datasets/question_answering.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    # into datafrane to decode string
+    df = tfds.as_dataframe(ds.take(max_examples))
+    df['context'] = df['context'].str.decode("utf-8")
+    df['question'] = df['question'].str.decode("utf-8")
+
+    # populate this with data records
+    self._examples = [{
+      'context': row['context'],
+      'question': row['question'],
+    } for _, row in df.iterrows()]
```
Comment (RyanMullins) [inline lit_nlp/examples/datasets/question_answering.py:24]: Given the impending changes to the `spec()` for this Dataset, we'll need to update this comprehension. Unfortunately, adding the other fields will violate the [Style Guide's rules for comprehensions](https://google.github.io/styleguide/pyguide.html#27-comprehensions--generator-expressions) (i.e., the mapping statement won't fit on one line), so we'll need to switch this to a for loop. 

```suggestion
    self._examples = []
    for _, row in df.iterrows():
      self._examples.append({
          # Field mapping code goes here
      })
```

File: lit_nlp/examples/tydi_demo.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+  models = {}
+  for model_name_or_path in _MODELS.value:
+    # Ignore path prefix, if using /path/to/<model_name> to load from a
+    # specific directory rather than the default shortcut.
+    model_name = os.path.basename(model_name_or_path)
+    models[model_name] = tydi.TyDiModel(model_name=model_name_or_path)
+
+  datasets = {
+      "tydi_qa": question_answering.TyDiQA(
+        split="validation-en", max_examples=_MAX_EXAMPLES.value),
```
Comment (RyanMullins) [inline lit_nlp/examples/tydi_demo.py:70]: Add additional entries for every split in the TyDi QA dataset and put the 2 character language code in the key names for this `dict`. Looks like you can just change the value of the `split` parameter and everything should work.

File: lit_nlp/examples/tydi_demo.py [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+
+  }
+
+  for name in datasets:
+    datasets[name] = datasets[name].slice[:_MAX_EXAMPLES.value]
```
Comment (RyanMullins) [inline lit_nlp/examples/tydi_demo.py:75]: I don't think you need to do this slice operation here as [`ds.take(...)`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#take) [does this for you](https://github.com/PAIR-code/lit/pull/753/files#diff-984c00264a68bced18d2649fefd20c1ba2e9a062236b4a7ba9a20d5e4b996a95R16) in the `TyDiQA` class constructor (when it converts the `tfds.Dataset` into a `pd.DataFrame`)

File: lit_nlp/examples/datasets/question_answering.py [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+  def spec(self) -> lit_types.Spec:
+    """Dataset spec, which should match the model"s input_spec()."""
+    return {
+        "context": lit_types.TextSegment(),
+        "question": lit_types.TextSegment(),
+    }
```
[the remark this answers is not in the document]
Comment (iftenney) [reply lit_nlp/examples/datasets/question_answering.py:31]: Something like this should work to construct the required `list[dtypes.AnnotationCluster]`:

```python
# answers is list[{'text': ..., 'start_char': ...}]
ret = []
for ans in answers:
  start = ans['start_char']
  end = ans['start_char'] + len(ans['text'])
  # Create a SpanLabel for this answer
  span = dtypes.SpanLabel(start, end, align='context')
  # Create a an AnnotationCluster containing this single answer span
  ret.append(dtypes.AnnotationCluster(label=ans['text'], spans=[span]) 

example['answers'] = ret
```

where in the data spec you have:

```python
  # exclusive=True means these are separate answer candidates, as opposed to e.g. a collection of entity tags
  "answers": lit_types.MultiSegmentAnnotations(exclusive=True)
```

File: lit_nlp/examples/models/tydi.py [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+
+        # creating output Dict
+        output = {
+            "generated_text" : self.tokenizer.decode(predict_answer_tokens),
+            # adding answer_text for debugging
+            "answers_text": answers_text[i]
+            # "tokens": self.tokenizer.convert_ids_to_tokens(tokenized_text.input_ids[i]),
+
+        }
+        prediction_output.append(output)
```
Comment (RyanMullins) [inline lit_nlp/examples/models/tydi.py:92]: You can do this all in one loop...

```suggestion
    prediction_output = []

    for inp in inputs:
        tokenized_text = self.tokenizer(inp['question'], inp['context'], return_tensors="jax",padding=True)
        results = self.model(**tokenized_text, output_attentions=True, output_hidden_states=True)

        answer_start_index = results.start_logits.argmax()
        answer_end_index = results.end_logits.argmax()
        predict_answer_tokens = tokenized_text.input_ids[0, answer_start_index : answer_end_index + 1]

        output = {
            "generated_text" : self.tokenizer.decode(predict_answer_tokens),
            # adding answer_text for debugging
            "answers_text": inp['answers_text']
            # "tokens": self.tokenizer.convert_ids_to_tokens(tokenized_text.input_ids[i]),
        }
        prediction_output.append(output)
```

File: lit_nlp/components/metrics.py [hunk 17]
```diff
@@ -226,6 +227,34 @@ def run_with_metadata(self,
     return self._metrics.run_with_metadata(indexed_inputs, model, dataset,
                                            model_outputs, margin_config)
 
+class ExactMatchMetrics(SimpleMetrics):
```
Comment (RyanMullins) [inline lit_nlp/components/metrics.py:230]: @aryan1107 please sync this branch to the latest on `dev`. My [latest change](https://github.com/PAIR-code/lit/commit/0d8341d9f120359bec86c983c5618dd59bb6f591) updates Metrics considerably and I expect there will be merge conflicts here that we'l need to resolve. Happy to pair program to resolve those.

File: lit_nlp/examples/models/tydi.py [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+
+  # Check outputs
+  ospec = model.output_spec()
+  assert "generated_text" in ospec
+  assert isinstance(
+      ospec["generated_text"],
+      (lit_types.GeneratedText))
+  assert ospec["generated_text"].parent == "answers_text"
+
+  return model
```
Comment (RyanMullins) [inline lit_nlp/examples/models/tydi.py:40]: Move this logic to tydi_test.py. 

The logic itself looks good enough for this demo, but since we don't use this function anywhere else inside this file, the logic probably belongs inside the test for this class.

T5 is an exception to this because the `SummarizationWrapper` and `TranslationWrapper` classes take an already initialized model as a parameter, and have to validate that model against a spec to ensure compatibility before doing the rest of their initialization. Since `TyDiModel` doesn't have any wrapper classes, we don't need to validate internally.
